### ISSUE TITLE
Request for help understanding the ledger cddl and provide test cases and examples

### ISSUE THREAD

**Issue by michele-nuzzi** [IntersectMBO/cardano-ledger#3313]
I'm trying to extract the `script_data_hash` of a transaction.

According to the [cddl](https://github.com/input-output-hk/cardano-ledger/blob/master/eras/babbage/test-suite/cddl-files/babbage.cddl#L94) the inputs is built as the `redeemers` field from the `tx_witness_set`; the `datums` field from the `tx_witness_set` and the `language_view`

In the `language_view` it is extremely unclear what is the representation of `language` and the representation of `script_integrity_data`

`language` seems to be a bytestring, even though in the `costmdls` cddl the key is a `uint`

`script_integrity_data` appears nowhere else in the cddl

I'm assuming it is the value of `costmdls` for that script's language, but is really not clear

in conclusion, it would be great to have examples to compare the cddl types with and possibly test cases, especially where hashing is required.

**Comment by JaredCorduan** [comment]
Thanks for bringing this up @michele-nuzzi , it would be great to make the script integrity hash clearer.

---

> `language` seems to be a bytestring, even though in the `costmdls` cddl the key is a `uint`

`language` is indeed a `uint`, as specified here:

https://github.com/input-output-hk/cardano-ledger/blob/f58187e1f2ebb7bbd9f5c3072cab5b1b6568c93f/eras/babbage/test-suite/cddl-files/babbage.cddl#L325-L326

with the caveat that in Plutus V1 we have this wart:

https://github.com/input-output-hk/cardano-ledger/blob/f58187e1f2ebb7bbd9f5c3072cab5b1b6568c93f/eras/babbage/test-suite/cddl-files/babbage.cddl#L129-L130

To make this concrete, the language tag for V1 is `4100`, and V2 is `01` (in CBOR hex).

---

> `script_integrity_data` appears nowhere else in the cddl
> 
> I'm assuming it is the value of `costmdls` for that script's language, but is really not clear

That is confusing indeed. We should make it clear in the prose that the _values_ in the map are the `script_integrity_data`. But your assumption is correct, the descriptions here are for the `script_integrity_data`:

https://github.com/input-output-hk/cardano-ledger/blob/f58187e1f2ebb7bbd9f5c3072cab5b1b6568c93f/eras/babbage/test-suite/cddl-files/babbage.cddl#L126-L132

To make this concrete, the `script_integrity_data` corresponding to the all zero costmodel for V1 would be encoded as:

```
58a89f00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000ff
```

and the `script_integrity_data` corresponding to the all zero costmodel for V2 would be encoded as:

```
98af0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```

---

> in conclusion, it would be great to have examples to compare the cddl types with and possibly test cases, especially where hashing is required.

I agree completely, examples would help a lot to bring clarity.

---

**To who ever wants to take this issue to make better documentation:**

* The encoding of the language views is here: https://github.com/input-output-hk/cardano-ledger/blob/f58187e1f2ebb7bbd9f5c3072cab5b1b6568c93f/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs#L658-L668
* we have some golden tests here: https://github.com/input-output-hk/cardano-ledger/blob/f58187e1f2ebb7bbd9f5c3072cab5b1b6568c93f/eras/alonzo/test-suite/test/Test/Cardano/Ledger/Alonzo/Golden.hs#L266-L296
